baseservice/base_service.py

```diff
@@ -0,0 +1,122 @@
+import logging
+import signal
+import threading
+from abc import ABCMeta, abstractmethod
+from enum import Enum, unique
+from typing import Optional
+
+from baseservice.utils import Event
+
+
+@unique
+class ServiceState(Enum):
+    STARTING = 'STARTING'
+    STOPPING = 'STOPPING'
+    STOPPED = 'STOPPED'
+
+
+class BaseService(metaclass=ABCMeta):
+    """
+    this class is the base class for all services
+    """
+
+    def __init__(self, *, name: str = None, should_stop_on_signal=True):
+        """
+
+        :param name: the name of this service. if None, the name of the type will be used
+        :param should_stop_on_signal: if True, the service will try to register SIGTERM and SIGINT on stop method.
+        """
+        self._should_stop_on_signal: bool = should_stop_on_signal
+        self._cancellation_token: threading.Event = threading.Event()
+        self._cancellation_token.set()  # service starts as not_running
+        self._logger = logging.getLogger(__name__)
+        self._state_event: Event[ServiceState] = Event()
+        if not name:
+            name = type(self).__name__
+
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def is_alive(self) -> bool:
+        """
+        this property is True, if the service is running.
+        """
+        return self._is_alive()
+
+    @property
+    def state_event(self) -> Event[ServiceState]:
+        """
+        this is an Event, that can be used to register on server state changes
+        """
+        return self._state_event
+
+    def _is_alive(self) -> bool:
+        """
+        this may be overridden by child classes to change the 'is_alive' behaviour
+        """
+        return not self._cancellation_token.is_set()
```

Comment:
No need to differentiate this from the usual is_alive.
It is already a property, so just move this logic to be directly there

Reply:
the only reason for this, is that properties are wierd while overriding... you need to remember to put the @property decoretor again, or it messes with your code. overriding a method is easier. but i wanted to make it a property. that's the solution i found